### PR TITLE
feat!: ble interface improvements

### PR DESCRIPTION
**Pull request by gabrielsantosphilips**
This pull request introduces the following improvements:

1. Added helper functions for streaming in GAP and GATT files.
2. Enhanced parser for advertising data, including local name and manufacturer specific data.
3. Simplified GATT client interface.
4. Moved Generic GAP definitions out from GapPeripheral classes for better organization.

### CODE REVIEW

File: services/ble/Gap.cpp [hunk 1]
```diff
@@ -96,4 +96,82 @@ namespace services
...
+        if (eventType == services::GapAdvertisingEventType::advInd)
+            stream << "ADV_IND";
+        else if (eventType == services::GapAdvertisingEventType::advDirectInd)
+            stream << "ADV_DIRECT_IND";
+        else if (eventType == services::GapAdvertisingEventType::advScanInd)
+            stream << "ADV_SCAN_IND";
+        else if (eventType == services::GapAdvertisingEventType::scanResponse)
+            stream << "SCAN_RESPONSE";
+        else
+        	stream << "ADV_NONCONN_IND";
```
[the remark this answers is not in the document]
Comment (gabrielsantosphilips) [reply services/ble/Gap.cpp:147]: Fixed.

File: services/ble/Gap.cpp [hunk 2]
```diff
@@ -96,4 +96,82 @@ namespace services
...
+    }
+
+    TextOutputStream& operator<<(TextOutputStream& stream, const services::GapAdvertisingEventAddressType& addressType)
+    {
+        if (addressType == services::GapAdvertisingEventAddressType::publicDeviceAddress)
+            stream << "Public Device Address";
+        else if (addressType == services::GapAdvertisingEventAddressType::randomDeviceAddress)
+            stream << "Random Device Address";
+        else
+        	stream << "Random Identity Address";
```
[the remark this answers is not in the document]
Comment (gabrielsantosphilips) [reply services/ble/Gap.cpp:159]: Fixed.

File: services/ble/test/TestGapCentral.cpp [hunk 3]
```diff
@@ -85,4 +86,46 @@ namespace services
         EXPECT_CALL(gap, StopDeviceDiscovery());
         decorator.StopDeviceDiscovery();
     }
+
+    TEST(GapAdvertisingDataParserTest, payload_does_not_contain_valid_info)
+    {
+        std::array<uint8_t, 1> data{{ 0x00 }};
```
[the remark this answers is not in the document]
Comment (gabrielsantosphilips) [reply services/ble/test/TestGapCentral.cpp:92]: Fixed.

File: services/ble/Gatt.cpp [hunk 4]
```diff
@@ -79,3 +79,40 @@ namespace services
         return 0;
     }
 }
+
+namespace infra
+{
+    TextOutputStream& operator<<(TextOutputStream& stream, const services::AttAttribute::Uuid& uuid)
+    {
+        if (uuid.Is<services::AttAttribute::Uuid16>())
+            stream << "0x" << hex << uuid.Get<services::AttAttribute::Uuid16>();
```
Comment (RoyCreemers) [inline services/ble/Gatt.cpp:88]: Why don't we use the same format (with square brackets) for both?

File: services/ble/Gap.cpp [hunk 5]
```diff
@@ -106,4 +106,86 @@ namespace services
     {
         GapCentralObserver::Subject().StopDeviceDiscovery();
     }
+
+    infra::ConstByteRange GapAdvertisingDataParser::LocalName(infra::ConstByteRange data)
+    {
+        auto localName = ParserAdvertisingData(data, GapAdvertisementDataType::shortenedLocalName);
```
Comment (RoyCreemers) [inline services/ble/Gap.cpp:112]: Swap shortened & full (cococo (and we) only use full, so saves an iteration

**Review by RoyCreemers — APPROVED** [review]

File: services/ble/Gatt.cpp (hunk 4 above)
Comment (gabrielsantosphilips) [reply services/ble/Gatt.cpp:88]: Good idea. Updated.

File: services/ble/Gap.cpp (hunk 5 above)
Comment (gabrielsantosphilips) [reply services/ble/Gap.cpp:112]: Done.

**Comment by EkelmansPh** [comment]
As indicated by SonarCloud, Gatt.cpp is missing some code coverage on new code.

File: services/ble/Gatt.cpp
Comment (EkelmansPh) [inline services/ble/Gatt.cpp:1]: As indicated by SonarCloud, Gatt.cpp is missing some code coverage on new code.

**Review by EkelmansPh — CHANGES_REQUESTED** [review]

File: services/ble/Gatt.cpp
Comment (gabrielsantosphilips) [reply services/ble/Gatt.cpp:1]: Fixed.

**Review by EkelmansPh — APPROVED** [review]

**Pull request merged**